tests/users/test_api_register.py

```diff
@@ -0,0 +1,178 @@
+        expected_tuple = [message, status_code]
+
+        http_error = requests.exceptions.HTTPError()
+        mock_response.raise_for_status.side_effect = http_error
+        mock_register.return_value = mock_response
+
+        data = {
+            "name": "user_pwd_invalid",
+            "username": "username_pwd_invalid",
+            "password": "toshort",
+            "email": "email_pwd_invalid",
+            "terms_and_conditions_checked": True,
+            "need_mentoring": True,
+            "available_to_mentor": False,
+        }
+
+        mock_error = Mock()
+        mock_error.json.return_value =  expected_tuple
+       
+        mock_register.side_effect = requests.exceptions.HTTPError(response=mock_error)
+        response = post_request(f"{BASE_MS_API_URL}/register", data)
+        mock_register.assert_called()
+        self.assertEqual(0, mock_response.raise_for_status.call_count)
+        self.assertEqual(0, mock_register.json.call_count)
+        self.assertEqual(0, mock_register.raise_for_status.call_count)
+        self.assertEqual(response[0], expected_tuple)
+        
+
+    @patch("requests.post")
+    def test_api_register_username_exist(self, mock_register):
+        mock_response = Mock()
+
+        json_response_code = json.dumps(HTTPStatus.CONFLICT)
+        message = {'message': '{messages.USER_USES_A_USERNAME_THAT_ALREADY_EXISTS}'}
+        status_code = {'status_code': f"{json_response_code}"}
+        expected_tuple = [message, status_code]
+
+        http_error = requests.exceptions.HTTPError()
+        mock_response.raise_for_status.side_effect = http_error
+        mock_register.return_value = mock_response
+        
+        data = {
+            "name": "user_username_exist",
+            "username": "username_exist",
+            "password": "pwd_username_exist",
+            "email": "email_username_exist",
+            "terms_and_conditions_checked": True,
+            "need_mentoring": True,
+            "available_to_mentor": False,
+        }
+
+        mock_error = Mock()
+        mock_error.json.return_value =  expected_tuple
+    
+        mock_register.side_effect = requests.exceptions.HTTPError(response=mock_error)
+        response = post_request(f"{BASE_MS_API_URL}/register", data)
+        mock_register.assert_called()
+        self.assertEqual(0, mock_response.raise_for_status.call_count)
+        self.assertEqual(0, mock_register.json.call_count)
+        self.assertEqual(0, mock_register.raise_for_status.call_count)
+        self.assertEqual(response[0], expected_tuple)
+
+        
+    @patch("requests.post")
+    def test_api_register_email_exist(self, mock_register):
+        mock_response = Mock()
+
+        json_response_code = json.dumps(HTTPStatus.CONFLICT)
+        message = {'message': '{messages.USER_USES_AN_EMAIL_ID_THAT_ALREADY_EXISTS}'}
+        status_code = {'status_code': f"{json_response_code}"}
+        expected_tuple = [message, status_code]
+
```

Comment:
Can you use the AAA to structure your tests? 
It's easy to read. A small change but will make your tests readable.
https://xp123.com/articles/3a-arrange-act-assert/

Reply:
I'm not following. I understand how it is done in frontend. But in Backend mocking API, can you show me how to rearrange this test so that it fits the 3A?

Reply:
Line 97 when you make the post request is your only action. So that's the middle, above that is all arrange, and below that is the assert block. Makes sense?

Reply:
so based on what you said, it is  already in the AAA structure, what else do you want me to change?

Reply:
do you mean to put a line gap for arrange, act and assert?

Reply:
do you mean rearrange the code like this?

<img width="777" alt="Screen Shot 2020-06-24 at 5 40 52 pm" src="https://user-images.githubusercontent.com/29667122/85516798-f4897500-b641-11ea-9e03-c7b7b0033f02.png">